File: quota/cards.py
"""Cards and display labels. Theme names sit beside suit marks."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

Suit = Literal["S", "H", "D", "C", "JOKER"]

SUITS: tuple[Suit, ...] = ("S", "H", "D", "C")

SUIT_MARK = {"S": "♠", "H": "♥", "D": "♦", "C": "♣", "JOKER": "Joker"}
SUIT_NAME = {"S": "香辛料", "H": "絹", "D": "宝石", "C": "茶", "JOKER": "銀貨"}

RANK_LABEL = {
    1: "A",
    11: "J",
    12: "Q",
    13: "K",
}


def bonus(rank: int) -> int:
    if rank <= 6:
        return 0
    if rank <= 9:
        return 1
    if rank <= 12:
        return 3
    if rank == 13:
        return 6
    raise ValueError(f"rank out of range: {rank}")


def score_for(rank: int) -> int:
    return rank + bonus(rank)


def sequence_bonus(cards: list[Card], same: int = 2, adjacent: int = 1) -> int:
    """Neighbor bonus. Jokers break the chain. K and A are not consecutive."""
    total = 0
    for left, right in zip(cards, cards[1:]):
        if left.suit == "JOKER" or right.suit == "JOKER":
            continue
        if left.rank == right.rank:
            total += same
        elif left.rank is not None and right.rank is not None and abs(left.rank - right.rank) == 1:
            total += adjacent
    return total


@dataclass(frozen=True, slots=True)
class Card:
    id: int
    suit: Suit
    rank: int | None

    def label(self) -> str:
        goods = SUIT_NAME[self.suit]
        if self.suit == "JOKER":
            return f"Joker 銀貨 #{self.id}"
        assert self.rank is not None
        face = RANK_LABEL.get(self.rank, str(self.rank))
        return f"{SUIT_MARK[self.suit]}{face} {goods} #{self.id}"


def make_deck(num_decks: int = 2) -> list[Card]:
    cards: list[Card] = []
    next_id = 0
    for _ in range(num_decks):
        for suit in SUITS:
            for rank in range(1, 14):
                cards.append(Card(next_id, suit, rank))
                next_id += 1
        for _j in range(2):
            cards.append(Card(next_id, "JOKER", None))
            next_id += 1
    return cards

File: unity/Assets/Scripts/Quota/Cards.cs
using System;
using System.Collections.Generic;

namespace Quota
{
    public enum Suit
    {
        S,
        H,
        D,
        C,
        Joker,
    }

    public sealed class Card
    {
        public readonly int Id;
        public readonly Suit Suit;
        public readonly int? Rank;

        public Card(int id, Suit suit, int? rank)
        {
            Id = id;
            Suit = suit;
            Rank = rank;
        }

        public string Label(ItemSet itemSet = null)
        {
            return (itemSet ?? ItemCatalog.Default()).Label(this);
        }
    }

    public static class Cards
    {
        public static readonly Suit[] KindSuits = { Suit.S, Suit.H, Suit.D, Suit.C };

        public static int Bonus(int rank)
        {
            if (rank <= 6) return 0;
            if (rank <= 9) return 1;
            if (rank <= 12) return 3;
            if (rank == 13) return 6;
            throw new ArgumentOutOfRangeException(nameof(rank), rank, "rank out of range");
        }

        public static int ScoreFor(int rank)
        {
            return rank + Bonus(rank);
        }

        public static int SequenceBonus(IReadOnlyList<Card> cards, int same = 2, int adjacent = 1)
        {
            var total = 0;
            for (var i = 1; i < cards.Count; i++)
            {
                var left = cards[i - 1];
                var right = cards[i];
                if (left.Suit == Suit.Joker || right.Suit == Suit.Joker) continue;
                if (left.Rank == right.Rank) total += same;
                else if (left.Rank != null && right.Rank != null && Math.Abs(left.Rank.Value - right.Rank.Value) == 1)
                    total += adjacent;
            }
            return total;
        }

        public static List<Card> MakeDeck(int numDecks = 2)
        {
            var cards = new List<Card>();
            var nextId = 0;
            for (var deck = 0; deck < numDecks; deck++)
            {
                foreach (var suit in KindSuits)
                {
                    for (var rank = 1; rank <= 13; rank++)
                        cards.Add(new Card(nextId++, suit, rank));
                }
                cards.Add(new Card(nextId++, Suit.Joker, null));
                cards.Add(new Card(nextId++, Suit.Joker, null));
            }
            return cards;
        }
    }
}
